File: backend/app/modules/learning/service.py
"""Учебная логика студента: доступность/завершение тем, попытки тестов и кейсов."""
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.modules.content.models import ClinicalCase, ContentItem, Question, Quiz
from app.modules.courses.models import Course, Enrollment
from app.modules.learning.models import CaseAttempt, Progress, QuizAttempt
from app.modules.learning.rules import date_gate_open, evaluate_completion, resolve_rule
from app.modules.topics.models import Topic


def is_enrolled(db: Session, student_id: int, course_id: int) -> bool:
    return db.scalar(
        select(Enrollment).where(
            Enrollment.student_id == student_id,
            Enrollment.course_id == course_id,
            Enrollment.status == "active",
        )
    ) is not None


def _published(db: Session, topic_id: int, kind: str) -> ContentItem | None:
    return db.scalar(
        select(ContentItem).where(
            ContentItem.topic_id == topic_id,
            ContentItem.kind == kind,
            ContentItem.status == "published",
        )
    )


def gather_facts(db: Session, student_id: int, topic: Topic) -> dict:
    """Собирает факты о прогрессе студента по теме для оценки правила."""
    progress = _get_progress(db, student_id, topic.id)

    quiz_item = _published(db, topic.id, "quiz")
    quiz_best, quiz_passed, has_quiz, quiz_attempted = 0.0, False, False, False
    if quiz_item:
        quiz = db.scalar(select(Quiz).where(Quiz.content_item_id == quiz_item.id))
        has_quiz = True
        attempts = db.scalars(
            select(QuizAttempt).where(
                QuizAttempt.quiz_id == quiz.id, QuizAttempt.student_id == student_id,
                QuizAttempt.finished_at.isnot(None),
            )
        ).all()
        if attempts:
            quiz_attempted = True
            quiz_best = max(a.score_pct for a in attempts)
            quiz_passed = any(a.passed for a in attempts)

    case_item = _published(db, topic.id, "case")
    has_case, case_submitted, case_reviewed = False, False, False
    if case_item:
        case = db.scalar(select(ClinicalCase).where(ClinicalCase.content_item_id == case_item.id))
        has_case = True
        ca = db.scalar(
            select(CaseAttempt).where(CaseAttempt.case_id == case.id, CaseAttempt.student_id == student_id)
        )
        if ca:
            case_submitted = True
            case_reviewed = ca.status == "reviewed"

    has_video = _published(db, topic.id, "video") is not None
    any_activity = (progress.video_watched_pct or 0) > 0 or quiz_attempted or case_submitted

    return {
        "has_video": has_video, "video_pct": progress.video_watched_pct,
        "has_quiz": has_quiz, "quiz_best_score": quiz_best, "quiz_passed": quiz_passed,
        "has_case": has_case, "case_submitted": case_submitted, "case_reviewed": case_reviewed,
        "any_activity": any_activity,
    }


def _get_progress(db: Session, student_id: int, topic_id: int) -> Progress:
    progress = db.scalar(
        select(Progress).where(Progress.student_id == student_id, Progress.topic_id == topic_id)
    )
    if progress is None:
        progress = Progress(student_id=student_id, topic_id=topic_id, state="locked")
        db.add(progress)
        db.flush()
    return progress


def recompute_course(db: Session, student_id: int, course_id: int) -> list[dict]:
    """Пересчитывает состояния всех тем курса для студента. Возвращает список
    {topic, state, video_pct, completed}."""
    course = db.get(Course, course_id)
    topics = db.scalars(
        select(Topic).where(Topic.course_id == course_id, Topic.status == "published")
        .order_by(Topic.order_index, Topic.id)
    ).all()

    result = []
    prev_completed = True  # для первой темы «предыдущая завершена» = True
    for topic in topics:
        rule = resolve_rule(topic.unlock_rule_json, course.settings_json)
        progress = _get_progress(db, student_id, topic.id)
        facts = gather_facts(db, student_id, topic)
        completed = evaluate_completion(rule, facts)

        available = (prev_completed and date_gate_open(rule)) or progress.manual_unlock
        if completed and available:
            state = "completed"
            if progress.completed_at is None:
                progress.completed_at = datetime.now(timezone.utc)
                # XP за завершение темы (идемпотентно по topic_id)
                from app.modules.gamification.service import award
                award(db, student_id, course_id, "lesson_completed", ref_id=topic.id)
        elif available:
            state = "in_progress" if facts["any_activity"] else "available"
        else:
            state = "locked"

        progress.state = state
        result.append({
            "topic": topic, "state": state,
            "video_pct": progress.video_watched_pct, "completed": completed,
        })
        prev_completed = completed and available
    db.commit()
    return result


def topic_state(db: Session, student_id: int, topic: Topic) -> str:
    """Состояние конкретной темы (пересчитывает курс — состояния зависят от предыдущих)."""
    states = recompute_course(db, student_id, topic.course_id)
    for row in states:
        if row["topic"].id == topic.id:
            return row["state"]
    return "locked"


def grade_quiz(db: Session, quiz: Quiz, answers: dict[int, int]) -> tuple[float, list[dict]]:
    """answers: {question_id: chosen_index}. Возвращает (score_pct, разбор по вопросам)."""
    questions = db.scalars(
        select(Question).where(Question.quiz_id == quiz.id).order_by(Question.order_index)
    ).all()
    if not questions:
        return 0.0, []
    correct = 0
    review = []
    for q in questions:
        chosen = answers.get(q.id, answers.get(str(q.id)))
        is_correct = chosen == q.correct_index
        if is_correct:
            correct += 1
        review.append({
            "question_id": q.id, "chosen": chosen, "correct_index": q.correct_index,
            "is_correct": is_correct,
        })
    return round(100.0 * correct / len(questions), 1), review

File: apps/api/src/modules/me/service.ts
import type { Prisma } from "../../lib/prisma";
import { prisma } from "../../lib/prisma";
import { ApiError, notFound } from "../../lib/errors";
import { DEFAULT_RULE, evaluateRule, lockedReason, type Facts, type Reason, type UnlockRule } from "./rules";

// A topic is visible to students only once it has at least one PUBLISHED content
// item. Un-published topics don't appear on the path at all.
const topicInclude = {
  contentItems: { where: { status: "PUBLISHED" }, include: { quiz: true, clinicalCase: true } },
} satisfies Prisma.TopicInclude;

// Faza 3: mavzular FANGA tegishli — kurs o'z fanining mavzularini ko'rsatadi.
const courseInclude = {
  subject: { include: { topics: { orderBy: { orderIndex: "asc" }, include: topicInclude } } },
  teacher: true,
  courseGroups: { include: { group: true } },
} satisfies Prisma.CourseInclude;

type CourseRow = Prisma.CourseGetPayload<{ include: typeof courseInclude }>;
type TopicWithContent = CourseRow["subject"]["topics"][number];
// Downstream kod (progress matritsa, tasks, lesson) `course.topics` bilan ishlaydi —
// loadCourse fan mavzularini shu maydonga normalizatsiya qiladi.
export type CourseWithTopics = CourseRow & { topics: TopicWithContent[] };

// Facts + extras: slides-viewed (shown on the card, doesn't gate) and a teacher
// manual-override flag that force-completes the topic (unlocks the next one).
export interface FullFacts extends Facts {
  slidesViewed: boolean;
  forceComplete: boolean;
}

function resolveRule(topic: TopicWithContent, course: CourseWithTopics): UnlockRule {
  const raw = (topic.unlockRuleJson ?? course.defaultUnlockRuleJson) as Partial<UnlockRule> | null;
  return { ...DEFAULT_RULE, ...(raw ?? {}) };
}

function buildElements(facts: FullFacts) {
  return {
    video: { exists: facts.hasVideo, watchedPct: facts.videoWatchedPct },
    slides: { exists: facts.hasSlides, viewed: facts.slidesViewed },
    quiz: { exists: facts.hasQuiz, score: facts.quizScore },
    case: { exists: facts.hasCase, submitted: facts.caseSubmitted, reviewed: facts.caseReviewed },
  };
}

export interface TopicOut {
  id: number;
  title: string;
  orderIndex: number;
  state: "LOCKED" | "AVAILABLE" | "IN_PROGRESS" | "COMPLETED";
  pct: number;
  reason: Reason | null;
  elements: ReturnType<typeof buildElements>;
}

/** The heart of the module: walk the topics in order, unlocking each only after
 *  the previous one is COMPLETED. Every LOCKED topic carries a concrete reason. */
export function computeTopics(course: CourseWithTopics, factsByTopic: Map<number, FullFacts>): TopicOut[] {
  const out: TopicOut[] = [];
  let prevCompleted = true; // the first topic is always open
  let prevUnmet: Reason[] = [];

  for (const topic of course.topics) {
    const rule = resolveRule(topic, course);
    const facts = factsByTopic.get(topic.id)!;
    const evaluated = evaluateRule(facts, rule);
    // A teacher override force-completes the topic regardless of the student's facts.
    const completed = evaluated.completed || facts.forceComplete;
    const pct = facts.forceComplete ? 100 : evaluated.pct;
    const dateOk = evaluated.dateOk || facts.forceComplete;
    const unmet = completed ? [] : evaluated.unmet;

    let state: TopicOut["state"];
    let reason: Reason | null = null;

    if (completed) {
      state = "COMPLETED";
    } else if (prevCompleted && dateOk) {
      state = pct > 0 ? "IN_PROGRESS" : "AVAILABLE";
    } else {
      state = "LOCKED";
      reason = lockedReason(rule, prevUnmet);
    }

    out.push({
      id: topic.id,
      title: topic.title,
      orderIndex: topic.orderIndex,
      state,
      pct,
      reason,
      elements: buildElements(facts),
    });

    prevCompleted = completed;
    prevUnmet = unmet;
  }

  return out;
}

/** Gather every fact the unlock engine needs for a student across a course's topics,
 *  in a few batched queries: progress rows, best quiz scores, case submissions. */
export async function studentFactsMap(studentId: number, course: CourseWithTopics): Promise<Map<number, FullFacts>> {
  const topicIds = course.topics.map((t) => t.id);
  const map = new Map<number, FullFacts>();
  if (topicIds.length === 0) return map;

  const progressRows = await prisma.progress.findMany({
    where: { studentId, topicId: { in: topicIds } },
    select: { topicId: true, videoWatchedPct: true, slidesViewed: true, overriddenAt: true },
  });
  const progressByTopic = new Map(progressRows.map((r) => [r.topicId, r]));

  // quizId -> topicId, caseId -> topicId
  const quizToTopic = new Map<number, number>();
  const caseToTopic = new Map<number, number>();
  for (const topic of course.topics) {
    for (const c of topic.contentItems) {
      if (c.quiz) quizToTopic.set(c.quiz.id, topic.id);
      if (c.clinicalCase) caseToTopic.set(c.clinicalCase.id, topic.id);
    }
  }

  // Best finished quiz score per topic.
  const bestScore = new Map<number, number>();
  if (quizToTopic.size > 0) {
    const attempts = await prisma.quizAttempt.findMany({
      where: { studentId, quizId: { in: [...quizToTopic.keys()] }, finishedAt: { not: null } },
      select: { quizId: true, scorePct: true },
    });
    for (const a of attempts) {
      const tid = quizToTopic.get(a.quizId)!;
      bestScore.set(tid, Math.max(bestScore.get(tid) ?? 0, a.scorePct));
    }
  }

  // Case submission / review per topic.
  const caseState = new Map<number, { submitted: boolean; reviewed: boolean }>();
  if (caseToTopic.size > 0) {
    const caseAttempts = await prisma.caseAttempt.findMany({
      where: { studentId, caseId: { in: [...caseToTopic.keys()] } },
      select: { caseId: true, reviewedAt: true },
    });
    for (const a of caseAttempts) {
      const tid = caseToTopic.get(a.caseId)!;
      caseState.set(tid, { submitted: true, reviewed: a.reviewedAt !== null });
    }
  }

  for (const topic of course.topics) {
    const kinds = new Set(topic.contentItems.map((c) => c.kind));
    const prog = progressByTopic.get(topic.id);
    const cs = caseState.get(topic.id);
    map.set(topic.id, {
      hasVideo: kinds.has("VIDEO"),
      hasSlides: kinds.has("PRESENTATION"),
      hasQuiz: kinds.has("QUIZ"),
      hasCase: kinds.has("CASE"),
      videoWatchedPct: prog?.videoWatchedPct ?? 0,
      slidesViewed: prog?.slidesViewed ?? false,
      forceComplete: prog?.overriddenAt != null,
      quizScore: bestScore.has(topic.id) ? bestScore.get(topic.id)! : null,
      caseSubmitted: cs?.submitted ?? false,
      caseReviewed: cs?.reviewed ?? false,
    });
  }

  return map;
}

function courseSummary(course: CourseWithTopics, topics: TopicOut[]) {
  const total = topics.length;
  const completed = topics.filter((t) => t.state === "COMPLETED").length;
  const next = topics.find((t) => t.state === "AVAILABLE" || t.state === "IN_PROGRESS") ?? null;
  return {
    id: course.id,
    subjectName: course.subject.name,
    teacherName: course.teacher.fullName,
    groupName: course.courseGroups[0]?.group.name ?? null,
    topicsTotal: total,
    topicsCompleted: completed,
    progressPct: total === 0 ? 0 : Math.round((completed / total) * 100),
    nextTopic: next?.title ?? null,
    nextTopicId: next?.id ?? null,
  };
}

export async function enrolledCourseIds(studentId: number): Promise<number[]> {
  const rows = await prisma.enrollment.findMany({
    where: { studentId, status: "ACTIVE" },
    select: { courseId: true },
    orderBy: { courseId: "asc" },
  });
  return rows.map((r) => r.courseId);
}

export async function loadCourse(courseId: number): Promise<CourseWithTopics> {
  const course = await prisma.course.findUnique({ where: { id: courseId }, include: courseInclude });
  if (!course) throw notFound("Kurs");
  // Only topics with at least one PUBLISHED content item are visible to students;
  // topics still being built (no published content) don't appear on the path at all.
  return { ...course, topics: course.subject.topics.filter((t) => t.contentItems.length > 0) };
}

/** GET /me/courses — enrolled courses with a progress summary each. */
export async function listMyCourses(studentId: number) {
  const ids = await enrolledCourseIds(studentId);
  const summaries = [];
  for (const id of ids) {
    const course = await loadCourse(id);
    const pm = await studentFactsMap(studentId, course);
    summaries.push(courseSummary(course, computeTopics(course, pm)));
  }
  return summaries;
}

/** GET /me/courses/:id — a course's full topic path for the student. */
export async function getMyCourse(studentId: number, courseId: number) {
  const enrolled = await prisma.enrollment.findUnique({
    where: { studentId_courseId: { studentId, courseId } },
  });
  if (!enrolled || enrolled.status !== "ACTIVE") {
    throw new ApiError(403, "forbidden", "Siz bu kursga yozilmagansiz", "Вы не записаны на этот курс");
  }
  const course = await loadCourse(courseId);
  const pm = await studentFactsMap(studentId, course);
  const topics = computeTopics(course, pm);
  return {
    id: course.id,
    subjectName: course.subject.name,
    teacherName: course.teacher.fullName,
    groupName: course.courseGroups[0]?.group.name ?? null,
    topicsTotal: topics.length,
    topicsCompleted: topics.filter((t) => t.state === "COMPLETED").length,
    progressPct: topics.length === 0 ? 0 : Math.round((topics.filter((t) => t.state === "COMPLETED").length / topics.length) * 100),
    topics,
  };
}

/** GET /me/dashboard — the "continue" block + course cards. */
export async function getDashboard(studentId: number) {
  const me = await prisma.user.findUnique({ where: { id: studentId }, select: { fullName: true } });
  const ids = await enrolledCourseIds(studentId);

  const courses = [];
  let resume: {
    courseId: number;
    subjectName: string;
    topicId: number;
    topic: string;
    pct: number;
  } | null = null;

  for (const id of ids) {
    const course = await loadCourse(id);
    const pm = await studentFactsMap(studentId, course);
    const topics = computeTopics(course, pm);
    courses.push(courseSummary(course, topics));

    if (!resume) {
      // Prefer an in-progress topic; otherwise the first available one.
      const current = topics.find((t) => t.state === "IN_PROGRESS") ?? topics.find((t) => t.state === "AVAILABLE");
      if (current) {
        resume = {
          courseId: course.id,
          subjectName: course.subject.name,
          topicId: current.id,
          topic: current.title,
          pct: current.pct,
        };
      }
    }
  }

  // Notifications: recently graded clinical cases ("your case was reviewed").
  const reviewed = await prisma.caseAttempt.findMany({
    where: { studentId, reviewedAt: { not: null } },
    orderBy: { reviewedAt: "desc" },
    take: 5,
    include: { clinicalCase: { include: { contentItem: { include: { topic: true } } } } },
  });
  const notifications = reviewed.map((a) => ({
    type: "case_reviewed" as const,
    caseAttemptId: a.id,
    topicId: a.clinicalCase.contentItem.topicId,
    topic: a.clinicalCase.contentItem.topic.title,
    score: a.score,
    reviewedAt: a.reviewedAt,
  }));

  return { fullName: me?.fullName ?? "", resume, courses, notifications };
}

// ---------- Shared access helpers (used by the lesson module) ----------

export function forbiddenNotEnrolled(): ApiError {
  return new ApiError(403, "forbidden", "Siz bu kursga yozilmagansiz", "Вы не записаны на этот курс");
}

export function forbiddenLocked(reason?: Reason): ApiError {
  return new ApiError(
    403,
    "topic_locked",
    reason?.uz ?? "Bu mavzu hali ochilmagan",
    reason?.ru ?? "Эта тема ещё не открыта"
  );
}

/** Faza 3: mavzu fanga tegishli — talabaning shu fandan ACTIVE yozilgan kursini topamiz. */
export async function enrolledCourseIdForTopic(studentId: number, topicId: number): Promise<number | null> {
  const topic = await prisma.topic.findUnique({ where: { id: topicId }, select: { subjectId: true } });
  if (!topic) throw notFound("Mavzu");
  const enrollment = await prisma.enrollment.findFirst({
    where: { studentId, status: "ACTIVE", course: { subjectId: topic.subjectId } },
    orderBy: { courseId: "asc" },
    select: { courseId: true },
  });
  return enrollment?.courseId ?? null;
}

/** Recomputes a single topic's state for the student, enforcing enrollment,
 *  published-visibility and the sequential lock. Throws 403/404 as appropriate. */
export async function assertTopicOpen(studentId: number, topicId: number): Promise<TopicOut> {
  const courseId = await enrolledCourseIdForTopic(studentId, topicId);
  if (courseId === null) throw forbiddenNotEnrolled();

  const course = await loadCourse(courseId);
  const facts = await studentFactsMap(studentId, course);
  const computed = computeTopics(course, facts);
  const state = computed.find((t) => t.id === topicId);
  if (!state) throw notFound("Mavzu"); // not published -> invisible to students
  if (state.state === "LOCKED") throw forbiddenLocked(state.reason ?? undefined);
  return state;
}

/** Recompute + return a single topic's fresh state (no lock enforcement) — used
 *  after a student action to report whether the topic just completed. */
export async function recomputeTopic(studentId: number, topicId: number): Promise<TopicOut | null> {
  const courseId = await enrolledCourseIdForTopic(studentId, topicId);
  if (courseId === null) return null;
  const course = await loadCourse(courseId);
  const facts = await studentFactsMap(studentId, course);
  return computeTopics(course, facts).find((t) => t.id === topicId) ?? null;
}

/** Recompute a topic's state and persist it to Progress (for the heatmap / completion
 *  stamp). Shared by student actions (lesson) and teacher grading (case review). */
export async function syncTopicProgress(studentId: number, topicId: number): Promise<TopicOut | null> {
  const t = await recomputeTopic(studentId, topicId);
  if (!t) return null;
  const row = await prisma.progress.upsert({
    where: { studentId_topicId: { studentId, topicId } },
    create: { studentId, topicId, state: t.state, completedAt: t.state === "COMPLETED" ? new Date() : null },
    update: { state: t.state },
  });
  if (t.state === "COMPLETED" && row.completedAt === null) {
    await prisma.progress.update({ where: { id: row.id }, data: { completedAt: new Date() } });
  }
  return t;
}
